# Question 1: 
Version of Pandas


In [1]:
import pandas as pd 
import numpy as np

In [2]:
pd.__version__

'2.1.0'

# Question 2: 
Number of columns in the dataset


In [31]:
df = pd.read_csv('raw.githubusercontent.com_alexeygrigorev_datasets_master_housing.csv')
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [4]:
df.shape

(20640, 10)

# Question 3: 
Select columns with missing data


In [35]:
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Answer: total_bedrooms        207

# Question 4: 
Number of unique values in the 'ocean_proximity' column


In [12]:
df['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [11]:
len(df['ocean_proximity'].unique())

5

# Question 5: 
Average value of the 'median_house_value' for the houses near the bay

In [19]:
avg_house_value =df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean()

In [20]:
avg_house_value

259212.31179039303

In [22]:
np.floor(avg_house_value)

259212.0

# Question 6: 
Has the mean value changed after filling missing values?


In [24]:
df.total_bedrooms.describe()

count    20433.000000
mean       537.870553
std        421.385070
min          1.000000
25%        296.000000
50%        435.000000
75%        647.000000
max       6445.000000
Name: total_bedrooms, dtype: float64

In [32]:
mean_total_bedrooms = np.round(df.total_bedrooms.mean(),3)
mean_total_bedrooms

537.871

In [36]:
df.total_bedrooms.fillna(mean_total_bedrooms,inplace=True)

In [37]:
df.total_bedrooms.mean()

537.8705570251938

In [39]:
np.round(df.total_bedrooms.mean(),3)

537.871

Still the same

In [38]:
df.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

# Questions 7: 
Value of the last element of w

In [46]:
#1
df_island = df[df.ocean_proximity == 'ISLAND']

In [49]:
#2
df_island_sub = df_island[['housing_median_age','total_rooms','total_bedrooms']]

In [50]:
df_island_sub

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [55]:
#3
X = df_island_sub.values
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [54]:
#4
X.T

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [58]:
X.T.dot(X)

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [59]:
XTX = X.T.dot(X)

In [60]:
np.linalg.inv(XTX)

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [62]:
XTX_inverse = np.linalg.inv(XTX)

In [61]:
#6
y = [950,1300,800,1000,1300]

In [63]:
#7 
XTX_inverse.dot(X.T).dot(y)

array([23.12330961, -1.48124183,  5.69922946])

In [64]:
w = XTX_inverse.dot(X.T).dot(y)

In [65]:
w[-1]

5.699229455065586